src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -193,39 +200,59 @@ public override void OnWriterCompleted(Action<Exception, object> callback, objec
         }
 
         /// <inheritdoc />
-        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        public override ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
         {
             // TODO ReadyAsync needs to throw if there are overlapping reads.
             ThrowIfCompleted();
 
-            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
-            CancellationTokenSource tokenSource = InternalTokenSource;
-            if (TryReadInternal(tokenSource, out ReadResult readResult))
+            if (TryReadInternal(out ReadResult readResult))
```

Comment:
Just noticed this. It wasn't even only TryRead that was affected 😨 Glad I caught this.

Reply:
an easy fix would be to avoid operations while there’s a pending read (return false from TryRead)

Reply:
That might be a reasonable short-term solution. It would be more correct to able to access previously unconsumed bytes if any, but I doubt being unable too would break many consumers. I guess we could file a followup issue.